pkg/operator/ceph/cluster/osd/spec.go

```diff
@@ -288,11 +292,9 @@ func (c *Cluster) makeDeployment(osdProps osdProperties, osd OSDInfo, provisionC
 	doBinaryCopyInit := true   // copy tini and rook binaries in an init container?
 	doActivateOSDInit := false // run an init container to activate the osd?
 
-	// If CVMode is empty, this likely means we upgraded Rook
-	// This property did not exist before so we need to initialize it
 	// This property is used for both PVC and non-PVC use case
 	if osd.CVMode == "" {
-		osd.CVMode = "lvm"
+		return nil, errors.Errorf("failed to generate deployment for OSD %d. required CVMode is not specified for this OSD", osd.ID)
 	}
```

Comment:
Return an error here since the code `getOSDInfo` sets `"lvm"` if this is empty. I ran into an issue updating a master-branch OSD that was created on node as "raw" because the `ROOK_CV_MODE` env var was unset for the new OSD. I also made changes to set this var on *all* OSD deployments now. I don't think we need to handle this with any special logic since `ROOK_CV_MODE=raw` was not backported for OSDs on nodes.

Reply:
The only time `CVMode` could be empty is when we are doing updates and have loaded OSDInfo from a deployment. But the `getOSDInfo` command sets `CVMode = "lvm"` if it is empty on a deployment like it always has. Otherwise, if it is a new OSD, that info is always returned from the provisioner. But if the provisioner doesn't return it due to some bug in the future, that should be an error.

Reply:
Ok but how can we tell if the error would be valid then? What if this is an upgrade?

Reply:
If this is an upgrade, `getOSDInfo()` reads the info from an existing deployment. If `CVMode` is not set on the deployment, it infers `"lvm"` and sets that on the `OSDInfo` struct returned from the function. This behavior of `getOSDInfo()` has not changed in my PR. IMO, we should not infer the OSD type in two places (both `getOSDInfo()` and here).

The change on the line here will now error in the following cases:
- the code from OSD prepare pods fails to report `CVMode` - IMO this is a good reason to error; new prepare pods should not be failing to provide this info
- the `getOSDInfo` code fails to set `CVMode` - IMO this is also a good reason to error; it means we are not properly reading info from existing deployments

The reason I changed this is because `CVMode` was not being set on updated OSD Deployments after your changes to add "raw" mode support on nodes. When a prepare pod created a new "raw" OSD, the deployment being created did not have `CVMode` populated. Because it was unpopulated, re-running the orchestration caused the "raw" OSD to be interpreted as "lvm" and the deployment update that followed resulted in a failing OSD. To fix the error I had to add `ROOK_CV_MODE` to the deployment. To prevent the error from happening again in the future, I believe it is good to error in this function if `CVMode` is unset so that we are not in the position again of ignoring a failure and causing more harm by doing so (i.e., an OSD that is updated in the wrong way and becomes broken).

Reply:
Ok thanks the clarification, I'm fine with returning an error in those circonstances.